server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -258,42 +254,38 @@ TopDocs mergeTopDocs(Collection<TopDocs> results, int topN, int from) {
             final TopFieldDocs[] shardTopDocs = results.toArray(new TopFieldDocs[numShards]);
 
             int sortLength = shardTopDocs[0].fields.length;
+            final String className = SortedNanosecondsNumericSortField.class.getName();
             for (int sortIdx = 0; sortIdx < sortLength; sortIdx++) {
                 final int idx = sortIdx;
 
                 boolean containsNanosecond = Arrays.stream(shardTopDocs)
                     .filter(d -> d.totalHits.value > 0)
-                    .map(d -> d.fields[idx].getClass().getName())
-                    .anyMatch(className -> SortedNanosecondsNumericSortField.class.getName().equals(className));
+                    .anyMatch(d -> className.equals(d.fields[idx].getClass().getName()));
 
                 // no nanoseconds, no need to do any work, check next sort index
                 if (containsNanosecond == false) {
                     continue;
                 }
 
                 boolean containsOnlyNanosecond  = Arrays.stream(shardTopDocs)
-                    .map(d -> d.fields[idx].getClass().getName())
-                    .allMatch(className -> SortedNanosecondsNumericSortField.class.getName().equals(className));
+                    .allMatch(d -> className.equals(d.fields[idx].getClass().getName()));
 
                 // only nanoseconds, no need to do any work, check next sort index
                 if (containsOnlyNanosecond) {
                     continue;
                 }
 
-                {
-                    List<String> sortFieldClasses = Arrays.stream(shardTopDocs)
-                        .map(d -> d.fields[idx])
-                        .map(sortField -> sortField.getClass().getName())
-                        .collect(Collectors.toList());
-                    logger.info("DO THE SORT # shards [{}] sortFieldClasses [{}]", shardTopDocs.length, sortFieldClasses);
-
-                }
                 // do the transformation for all shards
                 transformScoreToMilliseconds(shardTopDocs, sortIdx);
 
                 // only replace if the sort field is a nanosecond field, that is used by TopDocs.merge
                 if (sort.getSort()[sortIdx] instanceof SortedNanosecondsNumericSortField) {
-                    sort.getSort()[sortIdx] = ((SortedNanosecondsNumericSortField)sort.getSort()[sortIdx]).getSortField();
+                    SortedNumericSortField nsSortField = (SortedNumericSortField) sort.getSort()[sortIdx];
+                    SortField sortField = new SortedNumericSortField(nsSortField.getField(), nsSortField.getType(), nsSortField.getReverse(), nsSortField.getSelector());
```

Comment:
You don't need the selector when you merge the result, each document has a single sort value when we merge the top docs. So you can do:
````
SortField sortField = new SortField(nsSortField.getField(), nsSortField.getType(), nsSortField.getReverse());
````

Reply:
thanks! done